graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -169,3 +127,115 @@ def merge_schemas(schemas_dict):
+
+    Args:
+        query_type: str, name of the query type for the schema
+
+    Returns:
+        Document, representing a nearly blank schema
+    """
+    blank_ast = ast_types.Document(
+        definitions=[
+            ast_types.SchemaDefinition(
+                operation_types=[
+                    ast_types.OperationTypeDefinition(
+                        operation='query',
+                        type=ast_types.NamedType(
+                            name=ast_types.Name(value=query_type)
+                        ),
+                    )
+                ],
+                directives=[],
+            ),
+            ast_types.ObjectTypeDefinition(
+                name=ast_types.Name(value=query_type),
+                fields=[],
+                interfaces=[],
+                directives=[],
+            ),
+        ]
+    )
+    return blank_ast
+
+
+def _process_directive_definition(directive, existing_directives, merged_schema_ast):
+    """Compare new directive against existing directives, update records and schema.
+
+    Args:
+        directive: DirectiveDefinition, an AST node representing the definition of a directive
+        existing_directives: Dict[str, DirectiveDefinition], mapping the name of each existing
+                             directive to the AST node defining it; modified by this function
+        merged_schema_ast: Document, AST representing a schema; modified by this function
+    """
+    directive_name = directive.name.value
+    if directive_name in existing_directives:
+        if directive == existing_directives[directive_name]:
+            return
+        else:
+            raise SchemaNameConflictError(
+                u'Directive "{}" with definition "{}" has already been defined with '
+                u'definition "{}".'.format(
+                    directive_name,
+                    print_ast(directive),
+                    print_ast(existing_directives[directive_name]),
+                )
+            )
+    # new directive
+    merged_schema_ast.definitions.append(directive)
+    existing_directives[directive_name] = directive
+
+
+def _process_scalar_definition(scalar, existing_scalars, existing_types, merged_schema_ast):
+    """Compare new scalar against existing scalars and types, update records and schema.
+
+    Args:
+        scalar: ScalarDefinition, an AST node representing the definition of a scalar
+        existing_scalars: Set[str], set of names of all existing scalars; modified by this
+                          function
+        existing_types: Set[str], set of names of all existing types
+        merged_schema_ast: Document, AST representing a schema; modified by this function
+    """
+    scalar_name = scalar.name.value
+    if scalar_name in existing_scalars:
+        return
+    if scalar_name in existing_types:
+        raise SchemaNameConflictError(
+            u'New scalar "{}" clashes with existing type.'.format(scalar_name)
```

Comment:
This is an error message that is absolutely correct, and yet would be quite frustrating to address as a user. Two things would make it better: telling the user what they need to do about it, and giving them the necessary information such that they can do what needs doing.

For "what they need to do about it", I am worried that a user that sees this error might incorrectly conclude something like "oh, I guess I can't merge these schemas then." They might be entirely unfamiliar with the fact that the compiler already has a tool that is able to rename types -- and a subtle nudge here toward using that tool would ensure they know about it and react appropriately. One additional tricky thing: should they rename the scalar, or the other type? This is a good place to give users the advice to rename the other (non-scalar) type, since the scalar may come up e.g. in query output metadata as well, which could be more risky.

Given that they now want to rename the type, they need to find out which type from which schema is the problem. The error message never actually tells us either of these facts. Inspecting the code would tell them that the types conflict because they have the same name (which doesn't necessarily have to have been the case), but it still won't tell them which schema the type came from. And we might be merging dozens of schemas, each with many thousands of types.

Reply:
I've edited several of the name conflict error messages to include more information as well as suggestions to rename_schema. For the directive clashing error though, I don't keep track of what previous schema(s) the old definition came from, and rename_schema doesn't rename directives anyway, so I haven't enhanced the error message. Is that fine?

Reply:
It's probably okay for now, I think it's probably going to be a lot less common than conflicts between type names. If it becomes a problem down the line, we can fix it then.